crypto/rand/rand_win.c

```diff
@@ -20,8 +20,11 @@
 # include <windows.h>
 /* On Windows Vista or higher use BCrypt instead of the legacy CryptoAPI */
 # if defined(_MSC_VER) && _MSC_VER > 1500 /* 1500 = Visual Studio 2008 */ \
-     && defined(_WIN32_WINNT) && _WIN32_WINNT >= 0x0600
+     && defined(_WIN32_WINNT) && _WIN32_WINNT >= 0x0600 && defined(FIPS_MODE)
```

Comment:
no, this can't be right, the FIPS_MODE is in a provider.  I do not know how that is supposed to
work, bot.
others will know.

Reply:
The idea here is to:
 - keep USE_BCRYPTGENRANDOM only when building fips mode (and if _MSC_VER > 1500 && _WIN32_WINNT >= 0x0600)
 - try dynamic loading `USE_BCRYPTGENRANDOM_DL` otherwise (as long as it's not fips mode, which cannot use dynamic loading).

Reply:
the design of openssl 3.0 is no longer that way.
I think you need a PR for 1.1.1 and if you think that is necessary also a PR for 3.0
but your PR would be probably okay if you re-base for 1.1.1
and of course do a thorough test, I am not able to test anything with W in the name.

Reply:
> the design of openssl 3.0 is no longer that way.
> I think you need a PR for 1.1.1 and if you think that is necessary also a PR for 3.0
> but your PR would be probably okay if you re-base for 1.1.1
> and of course do a thorough test, I am not able to test anything with W in the name.

What does this mean: `with W in the name`? You mean Windows?
I run that change for half a year in hundreds of millions of installs, I specifically added the change because of sizable amount of users that are still using old OSes (in China for example).

Reply:
Oh, I did not say that, also vxWorks for instance has a W in the name,
maybe other O/S exist where I also have no access to the source code.

Reply:
All I said is that this PR cannot work in master but based on your
installed basis I will approve it immediately if you re-base it for 1.1.1

Reply:
@mspncp please have a look if my reaoning is correct.
Once the random number generator is moved to a provider it will
break this PR, but if @pps83 is using this PR on 10^8 Windows Installations
that make this an extemely urgent issue.  IT CANNOT WAIT.

Reply:
`FIPS_MODE` is meaningless in 1.1.1, it can never be defined.  I would not want to see this change going into 1.1.1 as is purely on this basis.

I'm not sure if dynamic loading is allowed in FIPS mode or not.  In 3.0, libcrypto will support it, the FIPS provider might not be allowed to -- I'm not sure.

I suggest that the legacy API be supported for versions that don't have BCrypt and that this isn't made conditional on FIPS.  Use of the legacy crypto API was changed in 2016 by fa64e63373fbc845a39907407ad990a6bbb84174 I think.

Reply:
The only reason I added that `FIPS_MODE` thing was because some builds in CI were failing or something like that